File: typescript/lexer/lexer.ts
import { Token, TokenType, lookupIdent } from "../token/token";

export class Lexer {
  input: string;

  position: number = 0; // current position in input (points to current char)
  readPosition: number = 0; // current reading position in input (after current char)
  ch: string | 0 = ""; // curent character under examination

  constructor(input: string) {
    this.input = input;
    this.readChar();
  }

  readChar() {
    if (this.readPosition >= this.input.length) {
      this.ch = 0;
    } else {
      this.ch = this.input.charAt(this.readPosition);
    }
    this.position = this.readPosition;
    this.readPosition += 1;
  }

  nextToken(): Token {
    let token: Token = { type: TokenType.ILLEGAL, literal: "" };
    this.skipWhitespace();

    switch (this.ch) {
      case "=":
        if (this.peekChar() === "=") {
          const ch = this.ch;
          this.readChar();
          const literal = ch + this.ch;
          token = newToken(TokenType.EQ, literal);
        } else {
          token = newToken(TokenType.ASSIGN, this.ch);
        }
        break;
      case "+":
        token = newToken(TokenType.PLUS, this.ch);
        break;
      case "-":
        token = newToken(TokenType.MINUS, this.ch);
        break;
      case "!":
        if (this.peekChar() === "=") {
          const ch = this.ch;
          this.readChar();
          const literal = ch + this.ch;
          token = newToken(TokenType.NOT_EQ, literal);
        } else {
          token = newToken(TokenType.BANG, this.ch);
        }
        break;
      case "/":
        token = newToken(TokenType.SLASH, this.ch);
        break;
      case "*":
        token = newToken(TokenType.ASTERISK, this.ch);
        break;
      case "<":
        token = newToken(TokenType.LT, this.ch);
        break;
      case ">":
        token = newToken(TokenType.GT, this.ch);
        break;
      case ";":
        token = newToken(TokenType.SEMICOLON, this.ch);
        break;
      case "(":
        token = newToken(TokenType.LPAREN, this.ch);
        break;
      case ")":
        token = newToken(TokenType.RPAREN, this.ch);
        break;
      case ",":
        token = newToken(TokenType.COMMA, this.ch);
        break;
      case "{":
        token = newToken(TokenType.LBRACE, this.ch);
        break;
      case "}":
        token = newToken(TokenType.RBRACE, this.ch);
        break;
      case 0:
        token = { type: TokenType.EOF, literal: "" };
        break;
      default:
        if (isLetter(this.ch)) {
          token.literal = this.readIdentifier();
          token.type = lookupIdent(token.literal);
          return token;
        } else if (isDigit(this.ch)) {
          token.type = TokenType.INT;
          token.literal = this.readNumber();
          return token;
        }
    }

    this.readChar();
    return token;
  }

  peekChar() {
    if (this.readPosition >= this.input.length) {
      return 0;
    } else {
      return this.input.charAt(this.readPosition);
    }
  }

  readIdentifier() {
    const position = this.position;
    while (isLetter(this.ch)) {
      this.readChar();
    }
    return this.input.slice(position, this.position);
  }

  readNumber() {
    const position = this.position;
    while (isDigit(this.ch)) {
      this.readChar();
    }
    return this.input.slice(position, this.position);
  }

  skipWhitespace() {
    while (
      this.ch === " " ||
      this.ch === "\t" ||
      this.ch === "\n" ||
      this.ch === "\r"
    ) {
      this.readChar();
    }
  }
}

function isLetter(ch: string | 0): boolean {
  if (ch === 0 || ch.length !== 1) return false;

  const code = ch.charCodeAt(0);
  return (
    (code >= 65 && code <= 90) || // A-Z
    (code >= 97 && code <= 122) || // a-z
    code === 95
  ); // underscore (_)
}

function isDigit(ch: string | 0): boolean {
  if (ch === 0 || ch.length !== 1) return false;
  const code = ch.charCodeAt(0);
  return code >= 48 && code <= 57;
}

function newToken(tokenType: TokenType, ch: string) {
  return new Token(tokenType, ch);
}

File: python/lexer/lexer.py
from types import NoneType
from monkey_token.token import Token, TokenType, lookup_ident


class Lexer:
    def __init__(self, input):
        self.input = input
        self.position = 0
        self.read_position = 0
        self.ch = ""
        self.read_char()

    def read_char(self):
        if self.read_position >= len(self.input):  # Use len() for string length
            self.ch = None  # Or use "" if you prefer an empty string to signify the end
        else:
            self.ch = self.input[self.read_position]
            self.position = self.read_position
            self.read_position += 1

    def next_token(self):
        token = None
        self.skip_whitespace()
        match self.ch:
            case "=":
                if self.peek_char() == "=":
                    ch = self.ch
                    self.read_char()
                    literal = ch + self.ch
                    token = Token(TokenType.EQ, literal)
                else:
                    token = Token(TokenType.ASSIGN, self.ch)
            case ";":
                token = Token(TokenType.SEMICOLON, self.ch)
            case "(":
                token = Token(TokenType.LPAREN, self.ch)
            case ")":
                token = Token(TokenType.RPAREN, self.ch)
            case ",":
                token = Token(TokenType.COMMA, self.ch)
            case "+":
                token = Token(TokenType.PLUS, self.ch)
            case "{":
                token = Token(TokenType.LBRACE, self.ch)
            case "}":
                token = Token(TokenType.RBRACE, self.ch)
            case "+":
                token = Token(TokenType.PLUS, self.ch)
            case "-":
                token = Token(TokenType.MINUS, self.ch)
            case "!":
                if self.peek_char() == "=":
                    ch = self.ch
                    self.read_char()
                    literal = ch + self.ch
                    token = Token(TokenType.NOT_EQ, literal)
                else:
                    token = Token(TokenType.BANG, self.ch)
            case "/":
                token = Token(TokenType.SLASH, self.ch)
            case "*":
                token = Token(TokenType.ASTERISK, self.ch)
            case "<":
                token = Token(TokenType.LT, self.ch)
            case ">":
                token = Token(TokenType.GT, self.ch)
            case None:
                token = Token(TokenType.EOF, "")
            case _:
                if is_letter(self.ch):
                    literal = self.read_identifier()
                    type = lookup_ident(literal)
                    return Token(type, literal)
                elif is_digit(self.ch):
                    literal = self.read_number()
                    return Token(TokenType.INT, literal)

        self.read_char()
        return token

    def read_identifier(self):
        position = self.position
        while is_letter(self.ch):
            self.read_char()
        return self.input[position : self.position]

    def read_number(self):
        position = self.position
        while is_digit(self.ch):
            self.read_char()
        return self.input[position : self.position]

    def skip_whitespace(self):
        while self.ch == " " or self.ch == "\t" or self.ch == "\n" or self.ch == "\r":
            self.read_char()

    def peek_char(self):
        if self.read_position >= len(self.input):
            return None
        else:
            return self.input[self.read_position]


def is_letter(ch: str | None):
    if not ch or len(ch) != 1:
        return False

    code = ord(ch)
    return (
        (code >= 65 and code <= 90)  # A-Z
        or (code >= 97 and code <= 122)  # a-z
        or code == 95  # underscore (_)
    )


def is_digit(ch: str | None):
    if not ch or len(ch) != 1:
        return False
    code = ord(ch)
    return code >= 48 and code <= 57
